state/protocol/badger/snapshot.go

```diff
@@ -41,6 +41,88 @@ func (s *Snapshot) Head() (*flow.Header, error) {
 	return head, err
 }
 
+// QuorumCertificate returns a valid quorum certificate for the block at the
+// head of this snapshot, if one exists.
+//
+// For root block snapshots, returns the root quorum certificate. For all other
+// blocks, generates a quorum certificate from a valid child, if one exists.
+func (s *Snapshot) QuorumCertificate() (*flow.QuorumCertificate, error) {
+
+	// CASE 1: for the root block, return the root QC
+	root, err := s.state.Params().Root()
+	if err != nil {
+		return nil, fmt.Errorf("could not get root: %w", err)
+	}
+
+	head, err := s.Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get head: %w", err)
+	}
+	if head.Height == root.Height {
+		// TODO store root QC and return here
+		return nil, fmt.Errorf("root qc not stored")
+	}
+
+	// CASE 2: for any other block, generate the root QC from a valid child
+	child, err := s.validChild()
+	if err != nil {
+		return nil, fmt.Errorf("could not get child: %w", err)
+	}
+
+	qc := &flow.QuorumCertificate{
+		View:      head.View,
+		BlockID:   child.ParentID,
+		SignerIDs: child.ParentVoterIDs,
+		SigData:   child.ParentVoterSig,
+	}
+
+	return qc, nil
+}
+
+// validChild returns a child of the snapshot head that has been validated
+// by HotStuff, if such a child exists. Otherwise returns an error. Any valid
+// child may be returned. Subsequent calls are not guaranteed to return the
+// same child.
+func (s *Snapshot) validChild() (*flow.Header, error) {
+
+	var childIDs []flow.Identifier
+	err := s.state.db.View(procedure.LookupBlockChildren(s.blockID, &childIDs))
+	if err != nil {
+		return nil, fmt.Errorf("could not look up children: %w", err)
+	}
+
+	// check we have at least one child
+	if len(childIDs) == 0 {
+		return nil, state.NewNoValidChildBlockError("block doesn't have children yet")
+	}
```

Comment:
Is this check necessary here? I feel that the extra cost of looping over an empty slice is negligible

Reply:
👍  Good point https://github.com/onflow/flow-go/pull/201/commits/ff6f97cf8f1e40bf9a7009541739908c6cb4b0ff